### PR TITLE
use memory mapped files for the peer buffer

### PR DESCRIPTION
**Pull request by arvidn**
to preserve state between restarts with some simple unit tests

### CODE REVIEW

File: mapped_file.hpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	size_t const m_size;
+};
+
+template <typename T>
+struct mapped_vector
+{
+	enum { header_size = 16 };
+	mapped_vector(char const* file, size_t const size)
+		: m_map(file, header_size + size * sizeof(T))
+		, m_size(*static_cast<size_t*>(m_map.data()))
```
Comment (ssiloti) [inline mapped_file.hpp]: This should be `reinterpret_cast` since you're transmuting the data rather than downcasting.

Comment (ssiloti) [inline mapped_file.hpp]: Never mind, I double checked the C++ spec and `static_cast` can indeed be used to convert void* to T*. Learn something new every day.

**Comment by arvidn** [comment]
one questionable thing, I think, is that the header size is not technically guaranteed to be a multiple of alignof(T) in mapped_vector. That should actually be pretty easy to fix.

File: main.cpp [hunk 2]
```diff
@@ -851,15 +870,8 @@ struct router_thread
...
-	void start(int threadid)
+	void start()
 	{
-		if (socks.empty())
-		{
-			fprintf(stderr, "no interfaces to receive on\n");
-			return;
-		}
-
-		this->threadid = threadid;
```
Comment (arvidn) [inline main.cpp]: I moved the thread id around a bit here, to have it be available in the constructor, in order to create unique names for the mapped files

File: main.cpp [hunk 3]
```diff
@@ -1247,12 +1259,6 @@ struct router_thread
 	int threadid;
 };
 
-void launch_router_thread(int threadid, std::vector<address> const& bind_addrs)
```
Comment (arvidn) [inline main.cpp]: I flattened this function into a lambda

File: main.cpp [hunk 4]
```diff
@@ -1466,15 +1489,16 @@ int main(int argc, char* argv[])
 		, std::ref(signals), std::ref(stats_timer), std::ref(ios)));
 
 	ios.run(ec);
+
+	for (auto& i : threads)
```
Comment (arvidn) [inline main.cpp]: the threads must be joined before destructed. the return below would cause a std::terminate in the thread destructors

**Comment by ssiloti** [comment]
lgtm, it would probably be a good idea to check the last modified time of the file and discard it if it is too old.

**Comment by arvidn** [comment]
Yeah, I was thinking about that too. I got a bit lazy and thought that the user could always just delete the file :)
maybe in another patch

**Comment by arvidn** [comment]
ok, fixed the alignment of T too

File: mapped_file.hpp [hunk 5]
```diff
@@ -86,7 +87,10 @@ struct mapped_file
 template <typename T>
 struct mapped_vector
 {
-	enum { header_size = 16 };
+	// the header must be large enough to make the first element still be
+	// correctly aligned
+	static constexpr size_t header_size = std::max(size_t(16), alignof(T));
```
Comment (ssiloti) [inline mapped_file.hpp]: std::max isn't constexpr in C++11, I don't think the qualifier is necessary in this case.

**Comment by ssiloti** [comment]
lgtm

**Comment by arvidn** [comment]
rebased on top of master again

**Pull request merged**